src/sparse/impl/KokkosSparse_par_ilut_numeric_impl.hpp

```diff
@@ -0,0 +1,1048 @@
+        [&](const size_type row, size_type& sum_inner) {
+          sum_inner += row_map(row);
+      }, sum);
+
+      team.team_barrier();
+
+      // Convert counts to local sums for this team. No team parallelism
+      // due to dependencies. Could potentially get some parallelism here
+      // by splitting into (endi - starti)/team_size chunks.
+      Kokkos::single(
+        Kokkos::PerTeam(team), [&] () {
+          prefix_sum_view(team_rank) = sum;
+          size_type curr_sum = 0;
+          for (size_type row = starti; row < endi; ++row) {
+            const size_type curr_val = row_map(row);
+            row_map(row) = curr_sum;
+            curr_sum += curr_val;
+          }
+      });
+  });
+
+  Kokkos::fence();
+
+  // Convert prefix_sum_view from per-team sums to global sums.
+  // This currently has no parallelism
+  Kokkos::parallel_for(
+    "prefix_sum per-team prefix",
+    RangePolicy(0, 1),
+    KOKKOS_LAMBDA(const size_type) {
+      size_type team_sum = 0;
+      for (size_type t = 0; t < nteams; ++t) {
+        const size_type team_result = prefix_sum_view(t);
+        prefix_sum_view(t) = team_sum;
+        team_sum += team_result;
+      }
+  });
+
+  Kokkos::fence();
+
+  // Add sums from earlier teams
+  size_type total_sum = 0;
+  Kokkos::parallel_reduce(
+    "prefix_sum finish",
+    policy,
+    KOKKOS_LAMBDA(const member_type& team, size_type& total_sum_arg) {
+      const auto team_rank = team.league_rank();
+      const size_type starti = team_rank * rows_per_team;
+      const size_type endi   = Kokkos::fmin(nrows, (team_rank + 1) * rows_per_team);
+      if (team_rank != 0) {
+        Kokkos::parallel_for(
+          Kokkos::TeamThreadRange(team, starti, endi),
+          [&](const size_type row) {
+            row_map(row) += prefix_sum_view(team_rank);
+        });
+      }
+
+      total_sum_arg = row_map(endi - 1);
+
+  }, Kokkos::Max<size_type>(total_sum));
+
+  Kokkos::fence();
+
+  return total_sum;
+}
+
+/**
+ * Just a convenience wrapper around spgemm
+ */
+template <class KHandle, class IlutHandle,
+          class LRowMapType, class LEntriesType, class LValuesType,
+          class URowMapType, class UEntriesType, class UValuesType,
+          class LURowMapType, class LUEntriesType, class LUValuesType>
+void multiply_matrices(
+  KHandle& kh, IlutHandle& ih,
+  const LRowMapType& L_row_map, const LEntriesType& L_entries, const LValuesType& L_values,
+  const URowMapType& U_row_map, const UEntriesType& U_entries, const UValuesType& U_values,
+  LURowMapType& LU_row_map,           LUEntriesType& LU_entries,     LUValuesType& LU_values)
+{
+  using execution_space = typename IlutHandle::execution_space;
+  using size_type       = typename IlutHandle::size_type;
+
+  const size_type nrows = ih.get_nrows();
+
+  KokkosSparse::Experimental::spgemm_symbolic(
+    &kh, nrows, nrows, nrows,
+    L_row_map, L_entries, false,
+    U_row_map, U_entries, false,
+    LU_row_map);
+
+  Kokkos::fence();
```

Comment:
Most of these fences can go away. When multiple kernels are launched on one execution space instance (which everything here is), they are run in sequence on the GPU. So even if kernel B depends on kernel A's results, you can still launch A and then B without a fence in between.

You only need a fence in a few situations I don't see here:
- you want a kernel to finish so you can measure its run time
- you are doing a reduction to a Kokkos::View, and you're going to read out of that view on host (reductions to view can be asynchronous, but reductions to plain values are naturally synchronous).
- you are going to compute something on the GPU in CudaUVMSpace (which is accessible on host and GPU), and then you want to read from that on host.